No API Generation with Webflux, Swagger 3.0.0-SNAPSHOT AND spring integration

I'm working with Spring Webflux, Swagger 3.0.0-SNAPSHOT and google cloud integration from spring (so Spring integration). I have several REST-Endpoints which needs to be documeted via swagger. I was following the normal implementation of swagger as described in https://github.com/springfox/springfox/issues/1773 . So far so good.

I ended with an almost empty swagger-ui page with following errors:
`😱 Could not render e, see the console.`
The Console Output was like:
`TypeError
columnNumber: 189990
fileName: "http://localhost:8080/webjars/springfox-swagger-ui/swagger-ui-standalone-preset.js?v=3.0.0-SNAPSHOT"
lineNumber: 13
message: "l[this.state.selectedIndex] is undefined"
stack: "value@http://localhost:8080/webjars/springfox-swagger-ui/swagger-ui-standalone-preset.js?v=3.0.0-SNAPSHOT:13:189990\nw/t.prototype.render@http://localhost:8080/webjars/springfox-swagger-ui/swagger-ui-bundle.js?v=3.0.0-SNAPSHOT:70:61143[etc.]`

The api-docs page just returned a 404 Not Found.
So the reason for the 404 is just an empty documentationCache in `Swagger2ControllerWebFlux`. I invested some breakpoint at startup, like in the loading class for all handler resolvers in `DocumentationPluginsBootstrapper` and no single method got called to populate the documentationCache.

The problem is the top annotation:
```
@Component
@Conditional(SpringIntegrationNotPresentInClassPathCondition.class)
public class DocumentationPluginsBootstrapper
```
The conditional clashes with the google cloud integration setup and leeds to an empty documentation for my webflux controllers. So since there is no documentation found, the api-docs page results in an 404 Not Found error and the swagger-ui page results in an not helpful javascript implementation error.

My workarround is to create the bean at startup on ym own like (kotlin-code):
```
@Bean
    fun swaggerDocumentationStarter(
            documentationPluginsManager: DocumentationPluginsManager,
            handlerProviders: List<RequestHandlerProvider>,
            scanned: DocumentationCache,
            resourceListing: ApiDocumentationScanner,
            typeResolver: TypeResolver,
            defaults: Defaults,
            pathProvider: PathProvider,
            environment: Environment): DocumentationPluginsBootstrapper = DocumentationPluginsBootstrapper(documentationPluginsManager, handlerProviders, scanned, resourceListing, typeResolver, defaults, pathProvider, environment)
```
This results in a total working swagger-ui and api-gen.

My question / change request would be to have a more apporiate solution for applications working with WebMvc / WebFlux AND spring integration.

BTW: Adding `springfox-spring-integration-webflux` as library results in following Stacktrace, so this sounds not like a solution:
```
11-04-2019 16:16:59.221 [main] ERROR o.s.boot.SpringApplication.reportFailure - Application run failed
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'springIntegrationWebFluxRequestHandlerProvider': Lookup method resolution failed; nested exception is java.lang.IllegalStateException: Failed to introspect Class [springfox.documentation.spring.web.plugins.SpringIntegrationWebFluxRequestHandlerProvider] from ClassLoader [sun.misc.Launcher$AppClassLoader@18b4aac2]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.determineCandidateConstructors(AutowiredAnnotationBeanPostProcessor.java:265)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineConstructorsFromBeanPostProcessors(AbstractAutowireCapableBeanFactory.java:1253)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1168)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:849)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:67)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
	at de.swr.ard.core.imageservice.service.ImageServiceApplicationKt.main(ImageServiceApplication.kt:87)
Caused by: java.lang.IllegalStateException: Failed to introspect Class [springfox.documentation.spring.web.plugins.SpringIntegrationWebFluxRequestHandlerProvider] from ClassLoader [sun.misc.Launcher$AppClassLoader@18b4aac2]
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:686)
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:583)
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:568)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.determineCandidateConstructors(AutowiredAnnotationBeanPostProcessor.java:248)
	... 18 common frames omitted
Caused by: java.lang.NoClassDefFoundError: org/springframework/integration/webflux/inbound/WebFluxIntegrationRequestMappingHandlerMapping
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.getDeclaredMethods(Class.java:1975)
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:668)
	... 21 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.springframework.integration.webflux.inbound.WebFluxIntegrationRequestMappingHandlerMapping
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 25 common frames omitted
```

--- a/springfox-spring-integration/src/main/java/springfox/documentation/spring/web/package-info.java
+++ b/springfox-spring-integration/src/main/java/springfox/documentation/spring/web/package-info.java
@@ -41,7 +41,7 @@
  * <p>
  * Changes in existing code:
  * <ul>
- * <li>{@code @Conditional(SpringIntegrationNotPresentInClassPathCondition.class)} on
+ * <li>{@code @Conditional(SpringIntegrationPluginNotPresentInClassPathCondition.class)} on
  * {@link springfox.documentation.spring.web.plugins.DocumentationPluginsBootstrapper} to avoid bootstrapping
  * the documentation plugins twice</li>
  * <li>Filter for Integration Handler Mappings in WebMvcRequestHandlerProvider and WebFluxRequestHandlerProvider</li>

--- a/springfox-spring-web/src/main/java/springfox/documentation/spring/web/plugins/DocumentationPluginsBootstrapper.java
+++ b/springfox-spring-web/src/main/java/springfox/documentation/spring/web/plugins/DocumentationPluginsBootstrapper.java
@@ -45,7 +45,7 @@ import java.util.concurrent.atomic.AtomicBoolean;
  * If no instances DocumentationConfigurer are found a default one is created and executed.
  */
 @Component
-@Conditional(SpringIntegrationNotPresentInClassPathCondition.class)
+@Conditional(SpringIntegrationPluginNotPresentInClassPathCondition.class)
 public class DocumentationPluginsBootstrapper
     extends AbstractDocumentationPluginsBootstrapper
     implements SmartLifecycle {

--- a/springfox-spring-web/src/main/java/springfox/documentation/spring/web/plugins/SpringIntegrationPluginNotPresentInClassPathCondition.java
+++ b/springfox-spring-web/src/main/java/springfox/documentation/spring/web/plugins/SpringIntegrationPluginNotPresentInClassPathCondition.java
@@ -25,11 +25,12 @@ import org.springframework.util.ClassUtils;
 
 import static org.springframework.util.ClassUtils.*;
 
-public class SpringIntegrationNotPresentInClassPathCondition implements Condition {
+public class SpringIntegrationPluginNotPresentInClassPathCondition implements Condition {
   @Override
   public boolean matches(ConditionContext context, AnnotatedTypeMetadata metadata) {
     return !isPresent(
-        "org.springframework.integration.handler.AbstractMessageHandler",
+        "springfox.documentation.spring.web.plugins" +
+            ".SpringIntegrationDocumentationPluginsBootstrapper",
         context.getClassLoader());
   }
 

